### PR TITLE
Add a null check for the endpoint on shutdown

### PR DESCRIPTION
**Pull request by kpayson64**
We are seeing a null pointer exception internally on the call to ```grpc_endpoint_shutdown()``` a few lines below internally.

Hypothetical race condition:

1: A handshaker completes successfully
2: ```on_handshake_done``` is _scheduled_ on the ```ExecCtx``` (with no error)
3:  ```grpc_handshake_manager_shutdown()``` is invoked on some other thread.
4: The handshakers shutdown function shuts down the endpoint and sets it to null (see rpc-switch)
5: ```ExecCtx::Flush()``` gets invoked on the original thread, ```on_handshake_done``` is invoked with ```GRPC_ERROR_NONE``` and ```mgr->shutdown=true```, and the endpoint has already been nulled out.

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                          FILE SIZE
 ++++++++++++++ GROWING                            ++++++++++++++
  +0.6%     +16 src/core/lib/channel/handshaker.cc     +16  +0.6%
      +1.9%     +16 call_next_handshaker_locked            +16  +1.9%

 -------------- SHRINKING                          --------------
  -0.0%     -16 [None]                                 -16  -0.0%

  [ = ]       0 TOTAL                                    0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                          FILE SIZE
 ++++++++++++++ GROWING                            ++++++++++++++
  +0.6%     +16 src/core/lib/channel/handshaker.cc     +16  +0.6%
      +1.9%     +16 call_next_handshaker_locked            +16  +1.9%

 -------------- SHRINKING                          --------------
  -0.0%     -16 [None]                                 -16  -0.0%

  [ = ]       0 TOTAL                                    0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

File: src/core/lib/channel/handshaker.cc [hunk 1]
```diff
@@ -223,18 +223,20 @@ static bool call_next_handshaker_locked(grpc_handshake_manager* mgr,
...
-      // point this can be removed.
-      grpc_endpoint_shutdown(mgr->args.endpoint, GRPC_ERROR_REF(error));
-      grpc_endpoint_destroy(mgr->args.endpoint);
-      mgr->args.endpoint = nullptr;
-      grpc_channel_args_destroy(mgr->args.args);
-      mgr->args.args = nullptr;
-      grpc_slice_buffer_destroy_internal(mgr->args.read_buffer);
-      gpr_free(mgr->args.read_buffer);
-      mgr->args.read_buffer = nullptr;
+      if (mgr->args.endpoint != nullptr) {
```
Comment (markdroth) [inline src/core/lib/channel/handshaker.cc]: Please add a comment explaining the race condition that we're catching here (the case you describe in the PR description).

**Comment by kpayson64** [comment]
I don't see a way to force this failure case with any end-to-end tests because the core surface has no guarantees about when flush gets called.

I think the only way to test this would be set up some kind of mock no-op handshaker and a unit test for the handshake manager.

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] Performance differences noted:
Benchmark                                                                                 cpu_time    real_time
----------------------------------------------------------------------------------------  ----------  -----------
BM_PumpStreamClientToServer<InProcess>/262144                                             +10%        +10%
BM_PumpStreamServerToClient<InProcess>/32768                                              +8%         +8%
BM_StreamingPingPong<InProcess, NoOpMutator, NoOpMutator>/262144/2                        +4%         +4%
BM_StreamingPingPongMsgs<InProcess, NoOpMutator, NoOpMutator>/262144                      +6%         +6%
BM_StreamingPingPongMsgs<InProcess, NoOpMutator, NoOpMutator>/32768                       +7%         +7%
BM_StreamingPingPongMsgs<MinInProcess, NoOpMutator, NoOpMutator>/32768                    +9%         +9%
BM_StreamingPingPongWithCoalescingApi<MinInProcess, NoOpMutator, NoOpMutator>/262144/2/1  +5%         +5%
BM_UnaryPingPong<MinInProcess, NoOpMutator, NoOpMutator>/262144/262144                    -4%         -4%
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                          FILE SIZE
 ++++++++++++++ GROWING                            ++++++++++++++
  +0.6%     +16 src/core/lib/channel/handshaker.cc     +16  +0.6%
      +1.9%     +16 call_next_handshaker_locked            +16  +1.9%

 -------------- SHRINKING                          --------------
  -0.0%     -16 [None]                                 -16  -0.0%

  [ = ]       0 TOTAL                                    0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by kpayson64** [comment]
#15693
#15751
#15752

**Pull request merged**